File: java/src/main/java/com/example/ddd/adapter/driving/RequestValueHelper.java
package com.example.ddd.adapter.driving;

import com.example.ddd.utility.JsonUtility;
import com.example.ddd.utility.ValueUtility;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.MissingNode;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

/**
 * 请求 Value Helper
 */
public class RequestValueHelper {

    /**
     * 获取请求体 json 数据
     */
    public static JsonNode toJson(String requestBody) {
        try {
            return JsonUtility.deserialize(requestBody);
        } catch (Exception e) {
            throw new ControllerException("请求体不是合法的JSON格式");
        }
    }

    //==================================================================================================================

    /**
     * 获取请求 json 数据中的值，失败返回 None
     */
    public static JsonNode getRequestJsonValue(JsonNode json, String... keys) {
        if (keys == null || keys.length == 0)
            return MissingNode.getInstance();
        JsonNode node = json;
        for (String key : keys) {
            node = node.path(key);
            if(node.isMissingNode())
                return MissingNode.getInstance();
        }
        return node;
    }

    /**
     * 获取请求 json 数据中的值，失败抛出异常
     */
    public static JsonNode getRequestJsonValueReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode()) {
            throw new ControllerException(String.format("请求体缺少 %s", String.join(".", keys)));
        }
        return node;
    }

    //==================================================================================================================

    /**
     * 获取请求 json 数据中 string 值，失败返回默认值
     */
    public static String getRequestJsonStringTrim(JsonNode json, String defaultValue, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode())
            return defaultValue;
        return node.asText().trim();
    }

    /**
     * 获取请求 json 数据中 string 值，失败返回空字符串
     */
    public static String getRequestJsonStringTrimOrEmpty(JsonNode json, String... keys) {
        return getRequestJsonStringTrim(json, "", keys);
    }

    /**
     * 获取请求 json 数据中 string 值，失败抛出异常
     */
    public static String getRequestJsonStringTrimReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValueReq(json, keys);
        return node.asText().trim();
    }

    //==================================================================================================================

    /**
     * 获取请求 json 数据中 bool 值，失败返回默认值
     */
    public static boolean getRequestJsonBool(JsonNode json, boolean defaultValue, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode())
            return defaultValue;
        if (node.isBoolean())
            return node.booleanValue();
        return ValueUtility.toBoolOrDefault(node.asText(), defaultValue);
    }

    /**
     * 获取请求 json 数据中 bool 值，失败抛出异常
     */
    public static boolean getRequestJsonBoolReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValueReq(json, keys);
        if (node.isBoolean())
            return node.booleanValue();
        Boolean b = ValueUtility.toBoolOrNull(node.asText());
        if (b == null)
            throw new ControllerException(String.format("请求体 %s 值不是合法的 bool", String.join(".", keys)));
        return b;
    }

    //==================================================================================================================

    /**
     * 获取请求 json 数据中 int 值，失败返回默认值
     */
    public static int getRequestJsonInt(JsonNode json, int defaultValue, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode())
            return defaultValue;
        if (node.isInt())
            return node.intValue();
        return ValueUtility.toIntOrDefault(node.asText(), defaultValue);
    }

    /**
     * 获取请求 json 数据中 int 值，失败抛出异常
     */
    public static int getRequestJsonIntReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValueReq(json, keys);
        if (node.isInt())
            return node.intValue();
        Integer i = ValueUtility.toIntOrNull(node.asText());
        if (i == null)
            throw new ControllerException(String.format("请求体 %s 值不是合法的 int", String.join(".", keys)));
        return i;
    }

    //==================================================================================================================

    /**
     * 获取请求 json 数据中 long 值，失败返回默认值
     */
    public static long getRequestJsonLong(JsonNode json, long defaultValue, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode())
            return defaultValue;
        if (node.isLong())
            return node.longValue();
        return ValueUtility.toLongOrDefault(node.asText(), defaultValue);
    }

    /**
     * 获取请求 json 数据中 long 值，失败抛出异常
     */
    public static long getRequestJsonLongReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValueReq(json, keys);
        if (node.isLong())
            return node.longValue();
        Long l = ValueUtility.toLongOrNull(node.asText());
        if (l == null)
            throw new ControllerException(String.format("请求体 %s 值不是合法的 long", String.join(".", keys)));
        return l;
    }

    //==================================================================================================================

    /**
     * 获取请求 json 数据中 datetime 值，失败返回默认值
     */
    public static LocalDateTime getRequestJsonDateTime(JsonNode json, LocalDateTime defaultValue, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode())
            return defaultValue;
        return ValueUtility.toDateTimeOrDefault(node.asText(), defaultValue);
    }

    /**
     * 获取请求 json 数据中 datetime 值，失败抛出异常
     */
    public static LocalDateTime getRequestJsonDateTimeReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValueReq(json, keys);
        LocalDateTime dt = ValueUtility.toDateTimeOrNull(node.asText());
        if (dt == null)
            throw new ControllerException(String.format("请求体 %s 值不是合法的 datetime", String.join(".", keys)));
        return dt;
    }

    //==================================================================================================================


    /**
     * 获取请求 json 数据中 date 值，失败返回默认值
     */
    public static LocalDate getRequestJsonDate(JsonNode json, LocalDate defaultValue, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode())
            return defaultValue;
        return ValueUtility.toDateOrDefault(node.asText(), defaultValue);
    }

    /**
     * 获取请求 json 数据中 date 值，失败抛出异常
     */
    public static LocalDate getRequestJsonDateReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValueReq(json, keys);
        LocalDate d = ValueUtility.toDateOrNull(node.asText());
        if (d == null)
            throw new ControllerException(String.format("请求体 %s 值不是合法的 date", String.join(".", keys)));
        return d;
    }

    //==================================================================================================================


    /**
     * 获取请求 json 数据中 time 值，失败返回默认值
     */
    public static LocalTime getRequestJsonTime(JsonNode json, LocalTime defaultValue, String... keys) {
        JsonNode node = getRequestJsonValue(json, keys);
        if (node.isMissingNode())
            return defaultValue;
        return ValueUtility.toTimeOrDefault(node.asText(), defaultValue);
    }

    /**
     * 获取请求 json 数据中 time 值，失败抛出异常
     */
    public static LocalTime getRequestJsonTimeReq(JsonNode json, String... keys) {
        JsonNode node = getRequestJsonValueReq(json, keys);
        LocalTime t = ValueUtility.toTimeOrNull(node.asText());
        if (t == null)
            throw new ControllerException(String.format("请求体 %s 值不是合法的 time", String.join(".", keys)));
        return t;
    }

    //==================================================================================================================

    /**
     * 分页信息
     */
    public record PagingInfo(int pageNum, int pageSize, int totalCount) { }

    /**
     * 验证并修正分页参数
     */
    public static PagingInfo fixPaging(int pageNum, int pageSize, int totalCount) {
        if (pageSize < 1)
            pageSize = 1;
        if (totalCount < 0)
            totalCount = 0;

        int maxPageNum = totalCount / pageSize;
        if (totalCount % pageSize > 0)
            maxPageNum++;

        if (pageNum > maxPageNum)
            pageNum = maxPageNum;
        if (pageNum < 1)
            pageNum = 1;

        return new PagingInfo(pageNum, pageSize, totalCount);
    }
}

File: dotnet/src/adapter/driving/RequestValueHelper.cs
using System.Text;
using System.Text.Json.Nodes;

using BackendExample.Utility;

namespace BackendExample.Adapter.Driving
{
    /// <summary>
    /// 请求 Value Helper
    /// </summary>
    public class RequestValueHelper
    {
        ///// <summary>
        ///// 获取请求体
        ///// </summary>
        //public static string GetRequestBody(HttpRequest request)
        //{
        //    using StreamReader reader = new StreamReader(request.Body, Encoding.UTF8);
        //    string body = reader.ReadToEnd();
        //    return body;
        //}

        /// <summary>
        /// 获取请求体，异步
        /// </summary>
        public static async Task<string> GetRequestBodyAsync(HttpRequest request)
        {
            using StreamReader reader = new StreamReader(request.Body, Encoding.UTF8);
            string body = await reader.ReadToEndAsync();
            return body;
        }

        ///// <summary>
        ///// 获取请求体 json 数据
        ///// </summary>
        //public static JsonNode GetRequestJson(HttpRequest request)
        //{
        //    try
        //    {
        //        return JsonUtility.Deserialize(GetRequestBody(request));
        //    }
        //    catch (Exception ex)
        //    {
        //        throw new ControllerException("请求体不是合法的JSON格式", ex);
        //    }
        //}

        /// <summary>
        /// 获取请求体 json 数据，异步
        /// </summary>
        public static async Task<JsonNode> GetRequestJsonAsync(HttpRequest request)
        {
            try
            {
                return JsonUtility.Deserialize(await GetRequestBodyAsync(request));
            }
            catch(Exception ex)
            {
                throw new ControllerException("请求体不是合法的JSON格式", ex);
            }
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中的值，失败返回 null
        /// </summary>
        public static JsonValue? GetRequestJsonValue(JsonNode json, params string[] keys)
        {
            if (keys == null || keys.Length == 0)
                return null;
            JsonNode? node = json;
            foreach (string key in keys)
            {
                node = json[key];
                if (node == null)
                    return null;
            }
            if (node is JsonValue value)
                return value;
            return null;
        }

        /// <summary>
        /// 获取请求 json 数据中的值，失败抛出异常
        /// </summary>
        public static JsonValue GetRequestJsonValueReq(JsonNode json, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                throw new ControllerException($"请求体缺少 {string.Join('.', keys)}");
            return value;
        }

        /// <summary>
        /// 获取请求 json 数据中的节点，失败返回 null
        /// </summary>
        public static JsonNode? GetRequestJsonNode(JsonNode json, params string[] keys)
        {
            if (keys == null || keys.Length == 0)
                return null;
            JsonNode? node = json;
            foreach (string key in keys)
            {
                node = json[key];
                if (node == null)
                    return null;
            }
            return node;
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 string 值，失败返回默认值
        /// </summary>
        public static string GetRequestJsonStringTrim(JsonNode json, string defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            if (value.TryGetValue<string>(out string? result) && result != null)
                return result.Trim();
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 string 值，失败返回空字符串
        /// </summary>
        public static string GetRequestJsonStringTrimOrEmpty(JsonNode json, params string[] keys)
        {
            return GetRequestJsonStringTrim(json, string.Empty, keys);
        }

        /// <summary>
        /// 获取请求 json 数据中 string 值，失败抛出异常
        /// </summary>
        public static string GetRequestJsonStringTrimReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            if (value.TryGetValue<string>(out string? result) && result != null)
                return result;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 string");
        }

        /// <summary>
        /// 获取请求 json 数据中 string 数组值
        /// </summary>
        public static IList<string> GetRequestJsonStringTrimList(JsonNode json, params string[] keys)
        {
            IList<string> ss = new List<string>();
            JsonNode? node = GetRequestJsonNode(json, keys);
            if (node != null && node is JsonArray a)
            {
                foreach (JsonNode? n in a)
                {
                    if (n != null && n is JsonValue v && v.TryGetValue<string>(out string? s))
                        ss.Add(s?.Trim() ?? string.Empty);
                }
            }
            return ss;
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 bool 值，失败返回默认值
        /// </summary>
        public static bool GetRequestJsonBool(JsonNode json, bool defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            if (value.TryGetValue<bool>(out bool boolResult))
                return boolResult;
            bool? b = null;
            if (value.TryGetValue<string>(out string? stringResult))
                b = ValueUtility.ToBoolOrNull(stringResult);
            if (b.HasValue)
                return b.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 bool 值，失败抛出异常
        /// </summary>
        public static bool GetRequestJsonBoolReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            if (value.TryGetValue<bool>(out bool boolResult))
                return boolResult;
            bool? b = null;
            if (value.TryGetValue<string>(out string? stringResult))
                b = ValueUtility.ToBoolOrNull(stringResult);
            if (b.HasValue)
                return b.Value;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 bool");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 int 值，失败返回默认值
        /// </summary>
        public static int GetRequestJsonInt(JsonNode json, int defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            if (value.TryGetValue<int>(out int intResult))
                return intResult;
            int? i = null;
            if (value.TryGetValue<string>(out string? stringResult))
                i = ValueUtility.ToIntOrNull(stringResult);
            if (i.HasValue)
                return i.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 int 值，失败抛出异常
        /// </summary>
        public static int GetRequestJsonIntReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            if (value.TryGetValue<int>(out int intResult))
                return intResult;
            int? i = null;
            if (value.TryGetValue<string>(out string? stringResult))
                i = ValueUtility.ToIntOrNull(stringResult);
            if (i.HasValue)
                return i.Value;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 int");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 long 值，失败返回默认值
        /// </summary>
        public static long GetRequestJsonLong(JsonNode json, long defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            if (value.TryGetValue<long>(out long longResult))
                return longResult;
            long? l = null;
            if (value.TryGetValue<string>(out string? stringResult))
                l = ValueUtility.ToLongOrNull(stringResult);
            if (l.HasValue)
                return l.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 long 值，失败抛出异常
        /// </summary>
        public static long GetRequestJsonLongReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            if (value.TryGetValue<long>(out long longResult))
                return longResult;
            long? l = null;
            if (value.TryGetValue<string>(out string? stringResult))
                l = ValueUtility.ToLongOrNull(stringResult);
            if (l.HasValue)
                return l.Value;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 long");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 decimal 值，失败返回默认值
        /// </summary>
        public static decimal GetRequestJsonDecimal(JsonNode json, decimal defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            if (value.TryGetValue<decimal>(out decimal decimalResult))
                return decimalResult;
            decimal? d = null;
            if (value.TryGetValue<string>(out string? stringResult))
                d = ValueUtility.ToDecimalOrNull(stringResult);
            if (d.HasValue)
                return d.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 decimal 值，失败抛出异常
        /// </summary>
        public static decimal GetRequestJsonDecimalReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            if (value.TryGetValue<decimal>(out decimal decimalResult))
                return decimalResult;
            decimal? d = null;
            if (value.TryGetValue<string>(out string? stringResult))
                d = ValueUtility.ToDecimalOrNull(stringResult);
            if (d.HasValue)
                return d.Value;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 decimal");
        }


        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 datetime 值，失败返回默认值
        /// </summary>
        public static DateTime GetRequestJsonDateTime(JsonNode json, DateTime defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            //if (value.TryGetValue<DateTime>(out DateTime dateTimeResult))
            //    return dateTimeResult;
            DateTime? dt = null;
            if (value.TryGetValue<string>(out string? stringResult))
                dt = ValueUtility.ToDateTimeOrNull(stringResult);
            if (dt.HasValue)
                return dt.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 datetime 值，失败抛出异常
        /// </summary>
        public static DateTime GetRequestJsonDateTimeReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            //if (value.TryGetValue<DateTime>(out DateTime dateTimeResult))
            //    return dateTimeResult;
            DateTime? dt = null;
            if (value.TryGetValue<string>(out string? stringResult))
                dt = ValueUtility.ToDateTimeOrNull(stringResult);
            if (dt.HasValue)
                return dt.Value;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 datetime");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 date 值，失败返回默认值
        /// </summary>
        public static DateOnly GetRequestJsonDate(JsonNode json, DateOnly defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            //if (value.TryGetValue<DateOnly>(out DateOnly dateResult))
            //    return dateResult;
            DateOnly? d = null;
            if (value.TryGetValue<string>(out string? stringResult))
                d = ValueUtility.ToDateOrNull(stringResult);
            if (d.HasValue)
                return d.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 date 值，失败抛出异常
        /// </summary>
        public static DateOnly GetRequestJsonDateReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            //if (value.TryGetValue<DateOnly>(out DateOnly dateResult))
            //    return dateResult;
            DateOnly? d = null;
            if (value.TryGetValue<string>(out string? stringResult))
                d = ValueUtility.ToDateOrNull(stringResult);
            if (d.HasValue)
                return d.Value;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 date");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求 json 数据中 time 值，失败返回默认值
        /// </summary>
        public static TimeOnly GetRequestJsonTime(JsonNode json, TimeOnly defaultValue, params string[] keys)
        {
            JsonValue? value = GetRequestJsonValue(json, keys);
            if (value == null)
                return defaultValue;
            //if (value.TryGetValue<TimeOnly>(out TimeOnly timeResult))
            //    return timeResult;
            TimeOnly? t = null;
            if (value.TryGetValue<string>(out string? stringResult))
                t = ValueUtility.ToTimeOrNull(stringResult);
            if (t.HasValue)
                return t.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求 json 数据中 time 值，失败抛出异常
        /// </summary>
        public static TimeOnly GetRequestJsonTimeReq(JsonNode json, params string[] keys)
        {
            JsonValue value = GetRequestJsonValueReq(json, keys);
            //if (value.TryGetValue<TimeOnly>(out TimeOnly timeResult))
            //    return timeResult;
            TimeOnly? t = null;
            if (value.TryGetValue<string>(out string? stringResult))
                t = ValueUtility.ToTimeOrNull(stringResult);
            if (t.HasValue)
                return t.Value;
            throw new ControllerException($"请求体 {string.Join('.', keys)} 值不是合法的 time");
        }


        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 string 值，失败返回默认值
        /// </summary>
        public static string GetRequestParamStringTrim(HttpRequest request, string defaultValue, string key)
        {
            string? value = request.Query[key];
            if (value == null)
                return defaultValue;
            return value;
        }

        /// <summary>
        /// 获取请求参数中 string 值，失败返回空字符串
        /// </summary>
        public static string GetRequestParamStringTrimOrEmpty(HttpRequest request, string key)
        {
            return GetRequestParamStringTrim(request, string.Empty, key);
        }

        /// <summary>
        /// 获取请求参数中 string 值，失败抛出异常
        /// </summary>
        public static string GetRequestParamStringTrimReq(HttpRequest request, string key)
        {
            string? value = request.Query[key];
            if (value == null)
                throw new ControllerException($"请求参数 {key} 值缺失");
            return value;
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 bool 值，失败返回默认值
        /// </summary>
        public static bool GetRequestParamBool(HttpRequest request, bool defaultValue, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            bool? b = ValueUtility.ToBoolOrNull(value);
            if (b.HasValue)
                return b.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求参数中 bool 值，失败抛出异常
        /// </summary>
        public static bool GetRequestParamBoolReq(HttpRequest request, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            bool? b = ValueUtility.ToBoolOrNull(value);
            if (b.HasValue)
                return b.Value;
            throw new ControllerException($"请求参数 {key} 值不是合法的 bool");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 int 值，失败返回默认值
        /// </summary>
        public static int GetRequestParamInt(HttpRequest request, int defaultValue, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            int? i = ValueUtility.ToIntOrNull(value);
            if (i.HasValue)
                return i.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求参数中 int 值，失败抛出异常
        /// </summary>
        public static int GetRequestParamIntReq(HttpRequest request, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            int? i = ValueUtility.ToIntOrNull(value);
            if (i.HasValue)
                return i.Value;
            throw new ControllerException($"请求参数 {key} 值不是合法的 int");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 long 值，失败返回默认值
        /// </summary>
        public static long GetRequestParamLong(HttpRequest request, long defaultValue, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            long? l = ValueUtility.ToLongOrNull(value);
            if (l.HasValue)
                return l.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求参数中 long 值，失败抛出异常
        /// </summary>
        public static long GetRequestParamLongReq(HttpRequest request, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            long? l = ValueUtility.ToLongOrNull(value);
            if (l.HasValue)
                return l.Value;
            throw new ControllerException($"请求参数 {key} 值不是合法的 long");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 decimal 值，失败返回默认值
        /// </summary>
        public static decimal GetRequestParamDecimal(HttpRequest request, decimal defaultValue, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            decimal? d = ValueUtility.ToDecimalOrNull(value);
            if (d.HasValue)
                return d.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求参数中 decimal 值，失败抛出异常
        /// </summary>
        public static decimal GetRequestParamDecimalReq(HttpRequest request, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            decimal? d = ValueUtility.ToDecimalOrNull(value);
            if (d.HasValue)
                return d.Value;
            throw new ControllerException($"请求参数 {key} 值不是合法的 decimal");
        }


        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 datetime 值，失败返回默认值
        /// </summary>
        public static DateTime GetRequestParamDateTime(HttpRequest request, DateTime defaultValue, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            DateTime? dt = ValueUtility.ToDateTimeOrNull(value);
            if (dt.HasValue)
                return dt.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求参数中 datetime 值，失败抛出异常
        /// </summary>
        public static DateTime GetRequestParamDateTimeReq(HttpRequest request, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            DateTime? dt = ValueUtility.ToDateTimeOrNull(value);
            if (dt.HasValue)
                return dt.Value;
            throw new ControllerException($"请求参数 {key} 值不是合法的 datetime");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 date 值，失败返回默认值
        /// </summary>
        public static DateOnly GetRequestParamDate(HttpRequest request, DateOnly defaultValue, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            DateOnly? d = ValueUtility.ToDateOrNull(value);
            if (d.HasValue)
                return d.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求参数中 date 值，失败抛出异常
        /// </summary>
        public static DateOnly GetRequestParamDateReq(HttpRequest request, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            DateOnly? d = ValueUtility.ToDateOrNull(value);
            if (d.HasValue)
                return d.Value;
            throw new ControllerException($"请求参数 {key} 值不是合法的 date");
        }

        //==============================================================================================================

        /// <summary>
        /// 获取请求参数中 time 值，失败返回默认值
        /// </summary>
        public static TimeOnly GetRequestParamTime(HttpRequest request, TimeOnly defaultValue, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            TimeOnly? t = ValueUtility.ToTimeOrNull(value);
            if (t.HasValue)
                return t.Value;
            return defaultValue;
        }

        /// <summary>
        /// 获取请求参数中 time 值，失败抛出异常
        /// </summary>
        public static TimeOnly GetRequestParamTimeReq(HttpRequest request, string key)
        {
            string value = GetRequestParamStringTrimOrEmpty(request, key);
            TimeOnly? t = ValueUtility.ToTimeOrNull(value);
            if (t.HasValue)
                return t.Value;
            throw new ControllerException($"请求参数 {key} 值不是合法的 time");
        }

        //==============================================================================================================

        /// <summary>
        /// 验证并修正分页参数
        /// </summary>
        public static (int pageNum, int pageSize, int totalCount) FixPaging(int pageNum, int pageSize, int totalCount)
        {
            if (pageSize < 1)
                pageSize = 1;
            if (totalCount < 0)
                totalCount = 0;

            int maxPageNum = totalCount / pageSize;
            if (totalCount % pageSize > 0)
                maxPageNum++;

            if (pageNum > maxPageNum)
                pageNum = maxPageNum;
            if (pageNum < 1)
                pageNum = 1;

            return (pageNum, pageSize, totalCount);
        }

    }
}
